pkg/db/check_statuses.go

```diff
@@ -29,3 +41,160 @@ func DeleteOldCheckStatuses() {
+		jobHistory.AddError(tx.Error.Error())
+		return
+	}
+
+	logger.Infof("Successfully deleted %v entries", tx.RowsAffected)
+	jobHistory.IncrSuccess()
+}
+
+// DeleteOld1hAggregatedCheckStatuses maintains retention period of old 1hour aggregated check statuses.
+func DeleteOld1hAggregatedCheckStatuses() {
+	jobHistory := models.NewJobHistory("DeleteOld1hAggregatedCheckStatuses", "", "").Start()
+	if err := PersistJobHistory(jobHistory); err != nil {
+		logger.Errorf("error persisting job history: %v", err)
+	}
+	defer func() {
+		if err := PersistJobHistory(jobHistory.End()); err != nil {
+			logger.Errorf("error persisting end of job: %v", err)
+		}
+	}()
+
+	const query = `DELETE FROM check_statuses_1h WHERE (NOW() - created_at) > INTERVAL '1 day' * ?`
+	tx := Gorm.Exec(query, RetentionDaysFor1hCheckStatusAggregate)
+	if tx.Error != nil {
+		logger.Errorf("error deleting old aggregated check statuses: %v", tx.Error)
+		jobHistory.AddError(tx.Error.Error())
+		return
+	}
+
+	logger.Infof("Successfully deleted %v entries", tx.RowsAffected)
+	jobHistory.IncrSuccess()
+}
+
+// AggregateCheckStatuses aggregates check statuses hourly
+// and stores it to the corresponding tables.
+func AggregateCheckStatuses1h() {
+	jobHistory := models.NewJobHistory("AggregateCheckStatuses1h", "", "").Start()
+	if err := PersistJobHistory(jobHistory); err != nil {
+		logger.Errorf("error persisting job history: %v", err)
+	}
+	defer func() {
+		if err := PersistJobHistory(jobHistory.End()); err != nil {
+			logger.Errorf("error persisting end of job: %v", err)
+		}
+	}()
+
+	if err := aggregateCheckStatuses(checkStatusAggDuration1h); err != nil {
+		logger.Errorf("error aggregating check statuses 1h: %v", err)
+		jobHistory.AddError(err.Error())
+	} else {
+		jobHistory.IncrSuccess()
+	}
+}
+
+// AggregateCheckStatuses aggregates check statuses daily
+// and stores it to the corresponding tables.
+func AggregateCheckStatuses1d() {
+	jobHistory := models.NewJobHistory("AggregateCheckStatuses1d", "", "").Start()
+	if err := PersistJobHistory(jobHistory); err != nil {
+		logger.Errorf("error persisting job history: %v", err)
+	}
+	defer func() {
+		if err := PersistJobHistory(jobHistory.End()); err != nil {
+			logger.Errorf("error persisting end of job: %v", err)
+		}
+	}()
+
+	if err := aggregateCheckStatuses(checkStatusAggDuration1d); err != nil {
+		logger.Errorf("error aggregating check statuses 1d: %v", err)
+		jobHistory.AddError(err.Error())
+	} else {
+		jobHistory.IncrSuccess()
+	}
+}
+
+func aggregateCheckStatuses(aggregateDurationType string) error {
+	const query = `
+		SELECT
+		check_statuses.check_id,
+		date_trunc(?, "time"),
+		count(*) AS total_checks,
+		count(*) FILTER (WHERE check_statuses.status = TRUE) AS passed,
+		count(*) FILTER (WHERE check_statuses.status = FALSE) AS failed,
+		SUM(duration) AS duration
+	FROM check_statuses 
+	LEFT JOIN checks ON check_statuses.check_id = checks.id
+	WHERE checks.created_at > NOW() - INTERVAL '1 day' * ?
```

Comment:
We should only be querying for the time range of 1h or 1d, and inserting the results - We shouldn't be return all the 1h values for the last day and then try upsert 24 values.

Reply:
But at the expense of doing 24x as much work and requiring a primary key - maybe just run it for the last 3h